File: src/core/agent/ticker-extract.ts
/** Token/ticker extraction — mirrors scripts/telegram-search/ticker_extract.py */

export const DEFAULT_TICKER_EXCLUDE = new Set([
	'THE',
	'AND',
	'FOR',
	'NOT',
	'YOU',
	'ALL',
	'NEW',
	'USD',
	'API',
	'CEO',
	'DAO',
	'TVL',
	'APR',
	'APY',
	'EVM',
	'NFT',
	'UTC',
	'GMT',
	'AI',
	'USDT',
	'USDC',
	'FAQ',
	'DM',
	'PM',
	'AM',
	'UK',
	'US',
	'EU',
]);

export type TickerDetectionConfig = {
	minLength: number;
	maxLength: number;
	includeBareCaps: boolean;
	exclude: Set<string>;
};

export const DEFAULT_TICKER_DETECTION: TickerDetectionConfig = {
	minLength: 2,
	maxLength: 10,
	includeBareCaps: true,
	exclude: DEFAULT_TICKER_EXCLUDE,
};

export function normalizeTicker(raw: string): string {
	return raw.trim().replace(/^[$#]+/, '').toUpperCase();
}

function symbolPattern(minLength: number, maxLength: number): string {
	const innerMin = Math.max(0, minLength - 1);
	const innerMax = Math.max(0, maxLength - 1);
	return `[A-Za-z][A-Za-z0-9]{${innerMin},${innerMax}}`;
}

export function extractTickers(
	text: string,
	config: TickerDetectionConfig = DEFAULT_TICKER_DETECTION,
): string[] {
	if (!text) {
		return [];
	}
	const sym = symbolPattern(config.minLength, config.maxLength);
	const found = new Set<string>();

	for (const match of text.matchAll(new RegExp(`\\$(${sym})`, 'g'))) {
		const token = match[1]!.toUpperCase();
		if (token.length >= config.minLength && token.length <= config.maxLength) {
			found.add(token);
		}
	}

	for (const match of text.matchAll(new RegExp(`#(${sym})`, 'g'))) {
		const token = match[1]!.toUpperCase();
		if (token.length >= config.minLength && token.length <= config.maxLength) {
			found.add(token);
		}
	}

	if (config.includeBareCaps) {
		for (const match of text.matchAll(new RegExp(`\\b(${sym})\\b`, 'g'))) {
			const token = match[1]!.toUpperCase();
			if (config.exclude.has(token)) {
				continue;
			}
			if (token.length >= config.minLength && token.length <= config.maxLength) {
				found.add(token);
			}
		}
	}

	return [...found].sort();
}

export function findTickersInMessage(
	text: string,
	allowlist: string[] | null,
	config: TickerDetectionConfig = DEFAULT_TICKER_DETECTION,
): string[] {
	if (allowlist && allowlist.length > 0) {
		const matched = new Set<string>();
		for (const raw of allowlist) {
			const sym = normalizeTicker(raw);
			if (!sym || config.exclude.has(sym)) {
				continue;
			}
			const patterns = [`\\$${sym}\\b`, `#${sym}\\b`];
			if (config.includeBareCaps) {
				patterns.push(`\\b${sym}\\b`);
			}
			for (const pattern of patterns) {
				if (new RegExp(pattern, 'i').test(text)) {
					matched.add(sym);
					break;
				}
			}
		}
		return [...matched].sort();
	}
	return extractTickers(text, config);
}

export function tickerDetectionFromYaml(raw: Record<string, unknown> | undefined): TickerDetectionConfig {
	if (!raw) {
		return DEFAULT_TICKER_DETECTION;
	}
	const excludeRaw = raw.exclude;
	const extraExclude = Array.isArray(excludeRaw)
		? excludeRaw.map((x) => String(x).trim().toUpperCase()).filter(Boolean)
		: [];
	const exclude = new Set([...DEFAULT_TICKER_EXCLUDE, ...extraExclude]);
	return {
		minLength: Number(raw.min_length ?? 2),
		maxLength: Number(raw.max_length ?? 10),
		includeBareCaps: raw.include_bare_caps !== false,
		exclude,
	};
}

export function tickerDetectionForPython(config: TickerDetectionConfig): Record<string, unknown> {
	return {
		min_length: config.minLength,
		max_length: config.maxLength,
		include_bare_caps: config.includeBareCaps,
		exclude: [...config.exclude],
	};
}

File: scripts/telegram-search/ticker_extract.py
"""Extract token/ticker symbols from Telegram message text."""

from __future__ import annotations

import re
from dataclasses import dataclass, field


DEFAULT_EXCLUDE = frozenset(
    {
        "THE",
        "AND",
        "FOR",
        "NOT",
        "YOU",
        "ALL",
        "NEW",
        "USD",
        "API",
        "CEO",
        "DAO",
        "TVL",
        "APR",
        "APY",
        "EVM",
        "NFT",
        "UTC",
        "GMT",
        "AI",
        "USDT",
        "USDC",
        "FAQ",
        "DM",
        "PM",
        "AM",
        "UK",
        "US",
        "EU",
    }
)


@dataclass(frozen=True)
class TickerDetectionConfig:
    min_length: int = 2
    max_length: int = 10
    include_bare_caps: bool = True
    exclude: frozenset[str] = DEFAULT_EXCLUDE


def normalize_ticker(raw: str) -> str:
    return raw.strip().lstrip("$#").upper()


def _symbol_pattern(min_length: int, max_length: int) -> str:
    inner_min = max(0, min_length - 1)
    inner_max = max(0, max_length - 1)
    return rf"[A-Za-z][A-Za-z0-9]{{{inner_min},{inner_max}}}"


def extract_tickers(text: str, config: TickerDetectionConfig) -> list[str]:
    if not text:
        return []
    sym = _symbol_pattern(config.min_length, config.max_length)
    found: set[str] = set()

    for match in re.finditer(rf"\$({sym})", text):
        token = match.group(1).upper()
        if config.min_length <= len(token) <= config.max_length:
            found.add(token)

    for match in re.finditer(rf"#({sym})", text):
        token = match.group(1).upper()
        if config.min_length <= len(token) <= config.max_length:
            found.add(token)

    if config.include_bare_caps:
        for match in re.finditer(rf"\b({sym})\b", text):
            token = match.group(1).upper()
            if token in config.exclude:
                continue
            if config.min_length <= len(token) <= config.max_length:
                found.add(token)

    return sorted(found)


def find_tickers_in_message(
    text: str,
    allowlist: list[str] | None,
    config: TickerDetectionConfig,
) -> list[str]:
    """Return matched tickers for one message. Empty allowlist = open detection."""
    if allowlist:
        matched: set[str] = set()
        for raw in allowlist:
            sym = normalize_ticker(raw)
            if not sym or sym in config.exclude:
                continue
            patterns = [
                rf"\${re.escape(sym)}\b",
                rf"#{re.escape(sym)}\b",
            ]
            if config.include_bare_caps:
                patterns.append(rf"\b{re.escape(sym)}\b")
            for pattern in patterns:
                if re.search(pattern, text, re.IGNORECASE):
                    matched.add(sym)
                    break
        return sorted(matched)

    return extract_tickers(text, config)


def config_from_dict(raw: dict | None) -> TickerDetectionConfig:
    if not raw:
        return TickerDetectionConfig()
    exclude_raw = raw.get("exclude") or []
    exclude = frozenset(
        {str(x).strip().upper() for x in exclude_raw if str(x).strip()}
    )
    if not exclude:
        exclude = DEFAULT_EXCLUDE
    else:
        exclude = DEFAULT_EXCLUDE | exclude
    return TickerDetectionConfig(
        min_length=int(raw.get("min_length", 2)),
        max_length=int(raw.get("max_length", 10)),
        include_bare_caps=bool(raw.get("include_bare_caps", True)),
        exclude=exclude,
    )
